src/main/java/com/google/javascript/clutz/ImportRenameMapBuilder.java

```diff
@@ -0,0 +1,128 @@
+
+    Node rightHandSide = destructuringAssignment.getChildAtIndex(1);
+
+    return rightHandSide.isCall()
+        && rightHandSide.getFirstChild().matchesQualifiedName("goog.require");
+  }
+
+  private String getGoogModuleId(Node astRoot) {
+    if (astRoot.getFirstChild() == null || !astRoot.getFirstChild().isModuleBody()) {
+      return null;
+    }
+
+    Node moduleBody = astRoot.getFirstChild();
+    for (Node statement : moduleBody.children()) {
+      if (isGoogModuleCall(statement)) {
+        return statement.getFirstChild().getChildAtIndex(1).getString();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * If we're running in incremental mode, closure gives imported symbols names in the current
+   * module's namespace, instead of the namespace of the module we're importing from. This reads the
+   * import statements to find the module id of the module that's being imported and builds a map
+   * from the names in the current module, to their correct counterparts.
+   */
+  private Map<String, String> buildImportRenameMap(String importingModuleId, Node astRoot) {
+    String abkljeri = astRoot.toStringTree();
+    if (astRoot.getFirstChild() == null || !astRoot.getFirstChild().isModuleBody()) {
+      return null;
+    }
+
+    Map<String, String> importRenameMap = new HashMap<>();
+    Node moduleBody = astRoot.getFirstChild();
+    for (Node statement : moduleBody.children()) {
+      if (isGoogRequireAssignment(statement)) {
+        String importedModuleId = statement.getFirstFirstChild().getChildAtIndex(1).getString();
+        String variableName = statement.getFirstChild().getString();
+        String symbolIdInCurrentModule =
+            "module$contents$" + importingModuleId.replace(".", "$") + "_" + variableName;
+        String symbolIdInDefiningModule = "module$exports$" + importedModuleId.replace(".", "$");
+        importRenameMap.put(symbolIdInCurrentModule, symbolIdInDefiningModule);
+      } else if (isGoogRequireDestructuringAssignment(statement)) {
+        String importedModuleId =
+            statement.getFirstChild().getChildAtIndex(1).getChildAtIndex(1).getString();
+        for (Node destructure : statement.getFirstFirstChild().children()) {
+          String originalName = destructure.getString();
+          // Destructuring can use the original name `const {A} = goog.require("foo.a")` or rename
+          // it `const {A: RenamedA} = ...`, and closure uses whichever in the symbole name it
+          // generates, so we have to extract it.
+          String variableName;
+          if (destructure.getFirstChild() != null) {
+            variableName = destructure.getFirstChild().getString();
+          } else {
+            variableName = originalName;
+          }
+          String symbolIdInCurrentModule =
+              "module$contents$" + importingModuleId.replace(".", "$") + "_" + variableName;
+          String symbolIdInDefiningModule =
```

Comment:
To expand on my comment about the required test: I think `"module$exports$bla"` is the symbol that's produced by a `goog.module`, but it is not the symbol that's produced by code using `goog.provide`. So if you have a `goog.module` that imports code using `goog.provide`, I think this bit here will to reference the right type.

The thing to test for is having multiple files that are compiled in incremental mode, and then piping the output to TypeScript compiler together.

Reply:
filed https://github.com/angular/clutz/issues/596

And agreed that we need tests for compiling the output of incremental mode.